parlai/core/tod/tod_core.py

```diff
@@ -0,0 +1,227 @@
+"""
+from enum import Enum
+from typing import List, Dict
+from dataclasses import dataclass, field
+from collections.abc import Iterable
+from parlai.utils.misc import warn_once
+
+STANDARD_CALL = "APICALL: "
+STANDARD_RESP = "APIRESP: "
+STANDARD_SYSTEM_UTTERANCE = "SYSTEM: "
+STANDARD_USER_UTTERANCE = "USER: "
+
+STANDARD_GOAL = "GOAL: "
+STANDARD_API_SCHEMAS = "APIS: "
+
+STANDARD_API_NAME_SLOT = "api_name"
+STANDARD_REQUIRED_KEY = "reqArg"
+STANDARD_OPTIONAL_KEY = "optArg"
+STANDARD_DONE = "[DONE]"
+
+CONST_SILENCE = "__SILENCE__"
+
+
+class TodAgentType(str, Enum):
+    USER_UTT_AGENT = "user_utt_model"
+    API_CALL_AGENT = "api_call_model"
+    API_RESP_AGENT = "api_resp_model"
+    SYSTEM_UTT_AGENT = "system_utt_model"
+    API_SCHEMA_GROUNDING_AGENT = "api_schema_grounding_model"
+    GOAL_GROUNDING_AGENT = "goal_grounding_model"
+
+
+TOD_AGENT_TYPE_TO_PREFIX = {
+    TodAgentType.USER_UTT_AGENT: STANDARD_USER_UTTERANCE,
+    TodAgentType.API_CALL_AGENT: STANDARD_CALL,
+    TodAgentType.API_RESP_AGENT: STANDARD_RESP,
+    TodAgentType.SYSTEM_UTT_AGENT: STANDARD_SYSTEM_UTTERANCE,
+    TodAgentType.API_SCHEMA_GROUNDING_AGENT: STANDARD_API_SCHEMAS,
+    TodAgentType.GOAL_GROUNDING_AGENT: STANDARD_GOAL,
+}
+
+
+@dataclass
+class TodStructuredRound:
+    """
+    Dataclass for rounds.
+    """
+
+    # Variables set by those using this class
+    user_utt: str = ""
+    api_call_machine: Dict = field(
+        default_factory=dict
+    )  # Hashmap of slot keys and slot values. Note that STANDARD_API_NAME_SLOT (`api_name`) is expected to be one of the keys here when this is nonempty; simulation metrics wonky without
+    api_resp_machine: Dict = field(default_factory=dict)
+    sys_utt: str = ""
+    extras: Dict = field(
+        default_factory=dict
+    )  # Grab bag for extra data. Not currently referenced in any TOD core code, but a convenient leaky abstraction for passing dataset-specific data between Parser classes and realized agents/teachers.
+
+    # Variables derived by class
+    api_call_utt: str = field(init=False)
+    api_resp_utt: str = field(init=False)
+
+    def __post_init__(self):
+        self.api_call_utt = SerializationHelpers.api_dict_to_str(self.api_call_machine)
+        self.api_resp_utt = SerializationHelpers.api_dict_to_str(self.api_resp_machine)
+        if (
+            len(self.api_call_machine) > 0
+            and STANDARD_API_NAME_SLOT not in self.api_call_machine
+        ):
+            warn_once(
+                f"{STANDARD_API_NAME_SLOT} missing when API Call present. This may cause issues for simulation metrics."
+            )
+
+
+@dataclass
+class TodStructuredEpisode:
+    """
+    Dataclass for episode-level data.
+    """
+
+    # Variables set by those using this class
+    delex: bool = False  # Set to true and this class will handle delexicalizing call + response utterances based on API calls and responses exposed to this class.
```

Comment:
Oof, need to find a linter that will wrap comments hmm...

Docformatter also fails to format the other comments inside of """'s. Playing around with it, not quite sure why (CC: @stephenroller )